lib/std/fs/File.zig

```diff
@@ -1206,12 +1206,33 @@ pub const Reader = struct {
         return r.size orelse {
             if (r.size_err) |err| return err;
             if (is_windows) {
-                if (windows.GetFileSizeEx(r.file.handle)) |size| {
-                    r.size = size;
-                    return size;
-                } else |err| {
-                    r.size_err = err;
-                    return err;
+                const device_type = blk: {
+                    var io_status: windows.IO_STATUS_BLOCK = undefined;
+                    var device_info: windows.FILE_FS_DEVICE_INFORMATION = undefined;
+                    const rc = windows.ntdll.NtQueryVolumeInformationFile(
+                        r.file.handle,
+                        &io_status,
+                        &device_info,
+                        @sizeOf(windows.FILE_FS_DEVICE_INFORMATION),
+                        .FileFsDeviceInformation,
+                    );
+                    break :blk switch (rc) {
+                        .SUCCESS => device_info.DeviceType,
+                        else => return windows.unexpectedStatus(rc),
+                    };
+                };
+                switch (device_type) {
+                    windows.FILE_DEVICE_NAMED_PIPE => {
```

Comment:
```suggestion
                    windows.FILE_DEVICE_NAMED_PIPE,
                    windows.FILE_DEVICE_CONSOLE,
                    => {
```

(confirmed that `FILE_DEVICE_CONSOLE` is returned for `stdin.handle` when not piping from a file/named pipe)

Could also include other device types, but this should cover the basics. Here's what reactos/wine do for reference:

- https://gitlab.winehq.org/wine/wine/-/blob/master/dlls/kernelbase/file.c#L3194-3200
- https://github.com/reactos/reactos/blob/050d8a8d65db5f9fd33c5e6779291f160822bf5e/dll/win32/kernel32/client/file/fileinfo.c#L308-L317
  + reactos seems to have some [special handling](https://github.com/reactos/reactos/blob/050d8a8d65db5f9fd33c5e6779291f160822bf5e/dll/win32/kernel32/client/file/fileinfo.c#L278-L283) for console handles that I'm failing to understand right now since [IsConsoleHandle](https://github.com/reactos/reactos/blob/050d8a8d65db5f9fd33c5e6779291f160822bf5e/sdk/lib/conutils/utils.c#L419) seems to call GetFileType...?